CBG-1143 Switch caching DCP feed to use gocb2

torcolvin opened the pull request:
- swap the lock system in change_cache for a channel, since
  DCPClient.Start() will start processing mutations before it is fully
  started. Add a Started channel to handle this in sgbucket

## Dependencies (if applicable)
- https://github.com/couchbase/sg-bucket/pull/76

## [Integration Tests]()
- [x] `xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/515/

bbrks commented on db/change_cache.go:
```diff
@@ -184,29 +183,34 @@ func (c *changeCache) Init(dbcontext *DatabaseContext, notifyChange func(base.Se
+		// TODO: how to test this error condition?
+		if err != nil {
+			base.ErrorfCtx(context.TODO(), "Failed to initialize task %w", err)
+			dbcontext.Close()
```
Does this need to return, to stop a nil `bgt` being added to the `c.backgroundTasks` slice? It seems we're in a state that we can't continue from, but we are continuing regardless.

torcolvin replied:
Added return for clarity.

bbrks commented on base/dcp_receiver.go:
```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
+	InfofCtx(loggingCtx, KeyDCP, "Started DCP Feed [%s]", MD(feedName))
+	go func() {
+		select {
+		case <-doneChan:
+			TracefCtx(loggingCtx, KeyDCP, "Closed DCP Feed [%s-%s].", MD(feedName))
+			err = <-doneChan
```
The 2nd read of `doneChan` is going to get a different value than the one read by the `case` clause

torcolvin replied:
Stopped shadowing this variable.

bbrks commented on base/dcp_receiver.go:
```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
+			close(args.DoneChan)
+		}
+	}()
+	return err
```
`err` is only ever assigned to in the above goroutine, which will have likely not started/run to completion before this function returns `err`

I'm not sure how we're handling the errors that come from closing the DCP client in this implementation.

bbrks commented on base/dcp_receiver.go:
```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
+
+	metadata := make([]DCPMetadata, numVbuckets)
+	for vbNo := uint16(0); vbNo < numVbuckets; vbNo++ {
+		// TODO: should we/can we initialize VbUUID or other fields?
```
I think so. Should be able to use the slice returned by `GetSTatsVbSeqno` with this index - although I have no idea about the failover entries...

torcolvin replied:
This is probably a question for @adamcfraser - these will be populated when the DCPClient starts, is my understanding.

bbrks commented on base/dcp_receiver.go:
```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
+		return err
+	}
+	bucketName := spec.BucketName
+	// TODO: does this need to be more unique, if two sync gateways connect to same server, will this reject the second DCP feed?
+	feedName := fmt.Sprintf("%s-%s:%s", args.ID, ProductAPIVersion, bucketName)
```
Yes, feed IDs must be unique or kv will terminate the connection of the existing client

We use `GenerateDcpStreamName` elsewhere to include a UUID in the name

markspolakovs replied:
OTOH, if this is for the caching feed / community import, will we expect to see multiple SGWs connecting?

torcolvin replied:
Changed to use common function.

bbrks commented on db/change_cache.go:
```diff
@@ -184,29 +183,34 @@ func (c *changeCache) Init(dbcontext *DatabaseContext, notifyChange func(base.Se
-		return err
-	}
-	c.backgroundTasks = append(c.backgroundTasks, bgt)
+	go func() {
+		dcpStartedErr := <-dcpStarted
```
As mentioned, let's discuss this change, as I'm not clear on what problem it's solving yet that the original lock isn't.

torcolvin replied:
This resolves the problem of mutation events occurring while `DCPClient.Start()` is happening, because that's blocking in `DatabaseContext.NewDatabaseContext`. Due to locking in the `changeCache`, we can deadlock here in `DCPClient.Start` which will send blocking events which are locked in `changeCache.processEntry`. We don't want housekeeping tasks to also lock in `changeCache.processEntry` while we are opening the database context, so we wait the database context is actually created.

This is a summary of an voice chat.

bbrks commented on db/change_cache.go:
```diff
@@ -184,29 +183,34 @@ func (c *changeCache) Init(dbcontext *DatabaseContext, notifyChange func(base.Se
-	defer c.lock.Unlock()
-
 	// Set initial sequence for sequence buffering
 	c._setInitialSequence(initialSequence)
```
The initial sequence is still set and read from places that are acquiring the lock (`DocChanged`, `changeCache.Clear`, `processEntry`, ...)

I'm not convinced this lock is safe to remove given the description of #3558

torcolvin replied:
Added back locking, since it will trigger the race detector in a test.

bbrks commented on db/change_cache_test.go:
```diff
@@ -1986,7 +1986,7 @@ func BenchmarkProcessEntry(b *testing.B) {
 
 			changeCache := &changeCache{}
-			if err := changeCache.Init(context, nil, nil); err != nil {
+			if err := changeCache.Init(context, nil, nil, nil); err != nil {
```
Passing in a `nil` channel here means the goroutine in Init will block forever trying to read from it, which doesn't seem correct.

torcolvin replied:
Check for `nil` channel here, for tests.

bbrks commented on base/dcp_client.go:
```diff
@@ -49,6 +49,7 @@ type DCPClientOptions struct {
 	FailOnRollback             bool           // When true, the DCP client will terminate on DCP rollback
 	InitialMetadata            []DCPMetadata  // When set, will be used as initial metadata for the DCP feed.  Will override any persisted metadata
 	CheckpointPersistFrequency *time.Duration // Overrides metadata persistence frequency - intended for test use
+	MetadataType               int            // Overrides metadata persistence frequency - intended for test use
```
- might be nice to have a type alias like `cbgtFeedType`
- comment needs updating

torcolvin replied:
Updated comments.

bbrks reviewed (changes requested):
first pass comments

bbrks commented:
I'm using this branch locally and I'm running into an issue where the caching feed does not start up correctly when I run SG, but also doesn't bubble any error or warning up to SG - resulting in a database running without being able to recieve any DCP mutations.

It seems to be reliably hitting this for whatever reason, but I'm not exactly sure why. Can you check if you see the same thing?

For one of my buckets (`b2` below) I see all 1024 vbuckets fail and the entire stream close with a trace log.
For another bucket, I see only a handful of vbuckets fail, and get a partial stream running.

```
2022-07-25T12:10:29.384+01:00 [INF] DCP: Starting mutation feed on bucket b2 due to either channel cache mode or doc tracking (auto-import)
2022-07-25T12:10:29.384+01:00 [INF] DCP: Using DCP feed for bucket: "b2" (based on feed_type specified in config file)
2022-07-25T12:10:29.468+01:00 [INF] DCP: Open stream for vbID 0 failed due to rollback or range error, will roll back metadata and retry: requested value is outside range | {"status_code":34,"bucket":"b2","error_name":"ERANGE","error_description":"Invalid range requested","opaque":14,"last_dispatched_to":"192.168.1.21:11210","last_dispatched_from":"192.168.1.19:64857","last_connection_id":"7c5415b81bd4e40d/db4b04e75ad7318f"}
2022-07-25T12:10:29.470+01:00 [DBG] DCP+: Stream (vb:0) closed, all items streamed
2022-07-25T12:10:29.472+01:00 [INF] DCP: Open stream for vbID 1 failed due to rollback or range error, will roll back metadata and retry: requested value is outside range | {"status_code":34,"bucket":"b2","error_name":"ERANGE","error_description":"Invalid range requested","opaque":16,"last_dispatched_to":"192.168.1.21:11210","last_dispatched_from":"192.168.1.19:64857","last_connection_id":"7c5415b81bd4e40d/db4b04e75ad7318f"}
2022-07-25T12:10:29.473+01:00 [DBG] DCP+: Stream (vb:1) closed, all items streamed
2022-07-25T12:10:29.474+01:00 [INF] DCP: Open stream for vbID 2 failed due to rollback or range error, will roll back metadata and retry: requested value is outside range | {"status_code":34,"bucket":"b2","error_name":"ERANGE","error_description":"Invalid range requested","opaque":18,"last_dispatched_to":"192.168.1.21:11210","last_dispatched_from":"192.168.1.19:64857","last_connection_id":"7c5415b81bd4e40d/db4b04e75ad7318f"}
2022-07-25T12:10:29.475+01:00 [DBG] DCP+: Stream (vb:2) closed, all items streamed

...

2022-07-25T12:10:31.792+01:00 [INF] DCP: Open stream for vbID 1023 failed due to rollback or range error, will roll back metadata and retry: requested value is outside range | {"status_code":34,"bucket":"b2","error_name":"ERANGE","error_description":"Invalid range requested","opaque":2060,"last_dispatched_to":"192.168.1.21:11210","last_dispatched_from":"192.168.1.19:64857","last_connection_id":"7c5415b81bd4e40d/db4b04e75ad7318f"}
2022-07-25T12:10:31.793+01:00 [DBG] DCP+: Stream (vb:1023) closed, all items streamed
2022-07-25T12:10:31.793+01:00 [INF] DCP: Started caching DCP Feed "SG-v-3.1-commit--uuid-6485b66e-0c0a-11ed-a616-acde48001122" for bucket "b2"
2022-07-25T12:10:31.793+01:00 [INF] Created background task: "InsertPendingEntries" with interval 2.5s
2022-07-25T12:10:31.793+01:00 [INF] Created background task: "CleanSkippedSequenceQueue" with interval 30m0s
2022-07-25T12:10:31.793+01:00 [TRC] DCP+: Closed caching DCP Feed "SG-v-3.1-commit--uuid-6485b66e-0c0a-11ed-a616-acde48001122" for "b2" -- base.StartGOCB2DCPFeed.func1() at dcp_receiver.go:279
```

torcolvin commented:
I've removed the locks that I added (I can remove the unused channel from walrus / sg-bucket. I'm not fully convinced that they don't exist but I can't reproduce this week and neither can jenkins.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+	metadata := make([]DCPMetadata, numVbuckets)
+	for vbNo := uint16(0); vbNo < numVbuckets; vbNo++ {
+		metadata[vbNo].VbUUID = gocbcore.VbUUID(vbUUIDs[vbNo])
+		metadata[vbNo].FailoverEntries = make([]gocbcore.FailoverEntry, 0)
```
Should we be initializing the FailoverEntries per vbucket, in the same way we're doing in makeVbucketMetadata?  (i.e. setting the VbUUID)

torcolvin replied:
I seeded this with VbUUID = 0, SeqNo = highSeqNo as per  https://github.com/couchbase/kv_engine/blob/master/docs/dcp/documentation/failure-scenarios.md#failover-log-behavior . The old behavior was to seed with SeqNo = 0

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// StartGOCB2DCPFeed starts a DCP Feed.
+func StartGOCB2DCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
```
It doesn't look like we're actually using BucketSpec here (could get spec.BucketName from bucket.Name()).  

If possible I think it would be preferable to pass in the parameters we actually need (including the collection set in future) instead of passing spec around.  In general I think the use case for BucketSpec should be 'I need to establish a new connection to the server from scratch'.

torcolvin replied:
This is a somewhat challenging problem to address right now as per offline discussion with @bbrks. We need the `Bucket` type in `DCPClient` if we are using persistent metadata. Moreover `Bucket` is actually a `Collection` which is conceptually wrong for a `DCPClient`, since it will be acting for multiple collections in a future code review, and we need actually pass something that can persist metadata for multiple collections, which oddly `Bucket` can not do.

I think we _do_ want to change this API, but not right now.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// StartGOCB2DCPFeed starts a DCP Feed.
+func StartGOCB2DCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
```
It looks like dbStats aren't being used.  At minimum I think we need to support the dcp_rollback_count stat, which I think is managed by the DCPClient today.  dcp_backfill_expected/dcp_backfill_completed are not needed.

torcolvin replied:
Added.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+		ErrorfCtx(loggingCtx, "Close error from caching DCP Feed %q for bucket %q: %w", feedName, MD(bucketName), closeErr)
+		return err
+	}
+	InfofCtx(loggingCtx, KeyDCP, "Started caching DCP Feed %q for bucket %q", feedName, MD(bucketName))
```
In the current DCP client we're only starting this goroutine if the terminator is non-nil.  Is the change to always run the goroutine intentional?

torcolvin replied:
I haven't found a case that's no-nil - I'd rather make it required if I could? We definitely terminate to shut the database down and the integration tests all pass.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+			// This is a close because DCP client closed on its own
+			InfofCtx(loggingCtx, KeyDCP, "Forced closed caching DCP Feed %q for %q", feedName, MD(bucketName))
+			// wait for channel close
+			<-doneChan
```
I don't think this is common practice (select on a channel, then block on the channel again inside the select).  It feels like a good way to end up in a deadlock that can't be escaped by closing the terminator.

One approach would be to just select on the channel, and then behave differently based on whether the error is nil.  The other would be to enhance the DCP client to not overload the done channel for both errors and done.

torcolvin replied:
Will fix in CBG-2234 soon after this PR goes through.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+			if dcpCloseError != nil {
+				WarnfCtx(loggingCtx, "Error on closing caching DCP Feed %q for %q: %w", feedName, MD(bucketName), dcpCloseError)
+			}
+			// FIXME: close dbContext here
```
Does the gocb v2 DCP client attempt to reconnect on error?  The cbdatasource DCP client does, and so SG doesn't need to build any of that handling.  If that's not the case for the new client, we'll need to at that either inside the DCP client, or here.

torcolvin replied:
Fixed this so that it reconnects indefinitely if this is not a one shot DCP feed after the initial connection (`DCPClient.Start`), which is retried for 5 minutes.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+			}
+			break
+		}
+		if args.DoneChan != nil {
```
Does it make sense to have two separate channels (args.DoneChan and the feed's doneChan)?

torcolvin replied:
Will fix in CBG-2234 soon after this PR goes through.

adamcfraser commented on base/dcp_mutation_feed.go:
```diff
@@ -0,0 +1,103 @@
+			if dcpCloseErr != nil {
+				WarnfCtx(loggingCtx, "Error on closing caching DCP Feed %q for %q: %w", feedName, MD(bucketName), dcpCloseErr)
+			}
+			dcpCloseErr = <-doneChan
```
I feel like it would be preferable to have dcpClient.Close() block on doneChan (or at least have an option to do that) - the double error handling here seems clunky.

torcolvin replied:
Will fix in CBG-2234 soon after this PR goes through.

adamcfraser reviewed (changes requested):
A few questions/comments.

torcolvin commented on base/bucket_gocb.go:
```diff
@@ -1500,7 +1500,7 @@ func (bucket *CouchbaseBucketGoCB) StartTapFeed(args sgbucket.FeedArguments, dbS
 
 	// TODO: Evaluate whether to use cbgt for non-sharded caching feed, as a way to push concurrency upstream
-	return StartDCPFeed(bucket, bucket.Spec, args, callback, dbStats)
+	return StartGOCB2DCPFeed(bucket, bucket.Spec, args, callback, dbStats)
```
I also moved and changed the messages here, because it's used as import feed for CE.

torcolvin reviewed (commented):
I filed future tickets for issues I didn't address in this code review. This has some intermittent failures in jenkins but I'm not sure that they are new or different from what we see.

bbrks commented on base/collection.go:
```diff
@@ -532,7 +532,11 @@ func (c *Collection) Incr(k string, amt, def uint64, exp uint32) (uint64, error)
 
 func (c *Collection) StartDCPFeed(args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
-	return StartDCPFeed(c, c.Spec, args, callback, dbStats)
+	if !IsEnterpriseEdition() {
+		// use new DCP feed in CBG-2232
+		return StartDCPFeed(c, c.Spec, args, callback, dbStats)
+	}
```
This does not feel like the right place to decide whether to use cbdatasource or gocb...

We _always_ want to use gocb for the mutation/changes feed, enterprise or not.

Can `StartImportFeed` skip calling `collection.StartDCPFeed` and instead call `Start(Cbdatasource)DCPFeed` directly? We have everything available up one level (including the EE/CE check), and import feels like the special case here.

![Screenshot 2022-08-02 at 14 09 52](https://user-images.githubusercontent.com/1525809/182382600-4a93b975-052b-475a-9e79-57bc1ce4f868.png)

markspolakovs replied:
In fact, I can't think of a reason why we'd need to use cbdatasource at all - community edition will still support collections (right?), so we have to use gocb for its import feed too. The decision of which feed to use for import "sharded or not sharded", not "gocb or cbdatasource" - could that help simplify this code?

torcolvin replied:
Moved this up a level.

bbrks reviewed (changes requested):
I'm ok with defferring the close cleanup work but how we choose whether to use cbdatasource/gocb needs fixing. See inline comment.
